server/data/data-jmap/src/main/java/org/apache/james/jmap/api/change/MailboxChange.java

```diff
@@ -92,60 +100,201 @@ public int getValue() {
+        }
+
+        public Builder destroyed(List<MailboxId> destroyed) {
+            this.destroyed = Optional.of(destroyed);
+            return this;
+        }
+
+        public MailboxChange build() {
+            return new MailboxChange(accountId, state, date, delegated, created.orElse(ImmutableList.of()), updated.orElse(ImmutableList.of()), destroyed.orElse(ImmutableList.of()));
+        }
+    }
+
+    public static RequiredAccountId builder() {
+        return accountId -> state -> date -> new Builder(accountId, state, date);
+    }
+
+    public static Builder created(AccountId accountId, State state, ZonedDateTime date, List<MailboxId> created) {
+        return MailboxChange.builder()
+            .accountId(accountId)
+            .state(state)
+            .date(date)
+            .created(created);
+    }
+
+    public static Builder updated(AccountId accountId, State state, ZonedDateTime date, List<MailboxId> updated) {
+        return MailboxChange.builder()
+            .accountId(accountId)
+            .state(state)
+            .date(date)
+            .updated(updated);
+    }
+
+    public static Builder destroyed(AccountId accountId, State state, ZonedDateTime date, List<MailboxId> destroyed) {
+        return MailboxChange.builder()
+            .accountId(accountId)
+            .state(state)
+            .date(date)
+            .destroyed(destroyed);
+    }
+
+    public static Optional<List<MailboxChange>> fromEvent(Event event, MailboxManager mailboxManager) {
         ZonedDateTime now = ZonedDateTime.now(Clock.systemUTC());
         if (event instanceof MailboxAdded) {
             MailboxAdded mailboxAdded = (MailboxAdded) event;
-            return Optional.of(MailboxChange.of(AccountId.fromUsername(mailboxAdded.getUsername()), State.of(UUID.randomUUID()), now, ImmutableList.of(mailboxAdded.getMailboxId()), ImmutableList.of(), ImmutableList.of()));
+            return Optional.of(ImmutableList.of(MailboxChange.created(AccountId.fromUsername(mailboxAdded.getUsername()), State.of(UUID.randomUUID()), now, ImmutableList.of(mailboxAdded.getMailboxId())).build()));
         }
         if (event instanceof MailboxRenamed) {
             MailboxRenamed mailboxRenamed = (MailboxRenamed) event;
-            return Optional.of(MailboxChange.of(AccountId.fromUsername(mailboxRenamed.getUsername()), State.of(UUID.randomUUID()), now, ImmutableList.of(), ImmutableList.of(mailboxRenamed.getMailboxId()), ImmutableList.of()));
+
+            MailboxChange ownerChange = MailboxChange.updated(AccountId.fromUsername(mailboxRenamed.getUsername()), State.of(UUID.randomUUID()), now, ImmutableList.of(mailboxRenamed.getMailboxId())).build();
+            Stream<MailboxChange> shareeChanges = getSharees(mailboxRenamed.getNewPath(), mailboxRenamed.getUsername(), mailboxManager)
+                .map(name -> MailboxChange.updated(AccountId.fromString(name), State.of(UUID.randomUUID()), now, ImmutableList.of(mailboxRenamed.getMailboxId()))
```

Comment:
`State.of(UUID.randomUUID())` -> `Satete.geneate()` ?

Hide your insides!

BTW with Cassandra relying on TimeUUID and not UUID itself, we will need a `State.Factory` interface for this `generate` method. With `State.DefaultFactory` calling UUID.randomUUID and `CassandraStateFactory` calling the time uuid version.